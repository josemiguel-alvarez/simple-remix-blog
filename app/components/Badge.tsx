import { Link } from "@remix-run/react";

interface Props {
  label: string;
  linkTo: string;
}

export const Badge = ({ label, linkTo }: Props) => {
  return (
    <div className="not-prose">
      <Link
        to={linkTo}
        className="no-underline relative inline-block text-sm font-medium text-blue-700 group"
      >
        <span className="absolute inset-0 transition-transform translate-x-0.5 translate-y-0.5 bg-blue-700 group-hover:translate-y-0 group-hover:translate-x-0"></span>

        <span className="relative block px-4 py-2 bg-white border border-current">
          {label}
        </span>
      </Link>
    </div>
  );
};
