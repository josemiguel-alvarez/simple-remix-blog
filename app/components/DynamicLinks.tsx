import { useMatches } from "@remix-run/react";
import type { AppData, LinkDescriptor } from "@remix-run/server-runtime";

export interface DynamicLinksFunction<Data extends AppData = AppData> {
  (args: { data: Data }): LinkDescriptor[];
}

interface DynamicLink {
  rel: string;
  href: string;
}

export function DynamicLinks() {
  const links: DynamicLink[] = useMatches().flatMap((match) => {
    let fn = match.handle?.dynamicLinks;
    if (typeof fn !== "function") return [];
    return fn({ data: match.data });
  });

  return (
    <>
      {links.map((link) => (
        <link key={link.rel} rel={link.rel} href={link.href} />
      ))}
    </>
  );
}
