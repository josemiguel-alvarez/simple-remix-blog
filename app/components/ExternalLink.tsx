interface Props {
  href: string;
  children: React.ReactNode;
}

export const ExternalLink = ({ href, children }: Props) => {
  return (
    <a href={href} target="_blank" rel="noopener noreferrer">
      {children}
    </a>
  );
};
