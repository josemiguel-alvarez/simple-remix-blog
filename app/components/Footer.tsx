import { Link } from "@remix-run/react";
import { siteMetadata } from "~/siteMetadata";
import { SocialMedia } from "./SocialMedia";

export const Footer = () => {
  return (
    <div>
      <SocialMedia />
      <div className="py-4 flex flex-col justify-center items-center space-x-2 text-xs sm:text-sm text-gray-500 dark:text-gray-400 sm:flex-row ">
        <div>{siteMetadata.author}</div>
        <div className="hidden sm:block">{` • `}</div>
        <div>{`© ${new Date().getFullYear()}`}</div>
        <div className="hidden sm:block">{` • `}</div>
        <Link to="/" className="no-underline text-gray-500">
          {siteMetadata.domain}
        </Link>
      </div>
    </div>
  );
};
