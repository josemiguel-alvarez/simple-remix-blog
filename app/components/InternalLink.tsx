import { Link } from "@remix-run/react";

interface Props {
  to: string;
  children: React.ReactNode;
}

export const InternalLink = ({ to, children }: Props) => {
  const beforeClasses =
    "before:content-[''] before:absolute before:block before:w-full before:h-0.5 before:bottom-0 before:left-0 before:bg-black dark:before:bg-white before:scale-x-0 before:transition-transform duration-300";

  return (
    <Link
      to={to}
      prefetch="intent"
      className={`relative no-underline w-fit ${beforeClasses} hover:before:scale-x-100`}
    >
      {children}
    </Link>
  );
};
