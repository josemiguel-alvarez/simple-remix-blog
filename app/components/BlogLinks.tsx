import { InternalLink } from "./InternalLink";

export const BlogLinks = () => {
  return (
    <>
      <InternalLink to="/blog">Blog</InternalLink>
      <InternalLink to="/tags">Tags</InternalLink>
      <InternalLink to="/about">About</InternalLink>
    </>
  );
};
