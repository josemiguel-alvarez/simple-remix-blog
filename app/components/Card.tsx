import type { Frontmatter } from "~/utils/posts.server";
import { InternalLink } from "./InternalLink";

type Props = Pick<Frontmatter, "title" | "summary" | "slug" | "formattedDate">;

export const Card = ({ title, summary, slug, formattedDate }: Props) => {
  return (
    <div>
      <div className="text-sm text-gray-500 dark:text-gray-400">
        Published {formattedDate}
      </div>
      <span className="text-2xl font-bold text-gray-800 dark:text-gray-200">
        {title}
      </span>
      <p className="text-lg text-gray-600 dark:text-gray-400">{summary}</p>
      <InternalLink to={`/posts/${slug}`}>
        <span className="font-medium text-black dark:text-white">
          Read more
        </span>
      </InternalLink>
    </div>
  );
};
