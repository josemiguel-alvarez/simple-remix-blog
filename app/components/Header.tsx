import Drawer from "@mui/material/Drawer";
import { Link } from "@remix-run/react";
import { useEffect, useState } from "react";
import { siteMetadata } from "~/siteMetadata";
import { isDarkMode } from "~/utils/darkMode";
import menuIcon from "../assets/menu.webp";
import { BlogLinks } from "./BlogLinks";

export const Header = () => {
  const [shouldShowDrawer, setShouldShowDrawer] = useState(false);
  const [siteLogo, setSiteLogo] = useState(siteMetadata.logo);

  const openDrawer = () => setShouldShowDrawer(true);
  const closeDrawer = () => setShouldShowDrawer(false);

  useEffect(() => {
    if (isDarkMode() && siteMetadata.logo_dark_mode) {
      setSiteLogo(siteMetadata.logo_dark_mode);
    }
  }, []);

  return (
    <header className="flex justify-between items-center max-w-full w-full py-8 gap-x-12 md:gap-x-0">
      <Link
        className="home text-3xl font-medium no-underline flex-1 m-0 not-prose md:my-4"
        to="/"
      >
        {siteLogo ? (
          <img alt="Website logo" src={siteLogo} loading="lazy" />
        ) : (
          <span>{siteMetadata.domain}</span>
        )}
      </Link>

      <div className="sm:flex items-center gap-4 hidden flex-1 justify-end">
        <BlogLinks />
      </div>

      <div className="sm:hidden not-prose" onClick={openDrawer}>
        <img
          alt="Menu"
          src={menuIcon}
          width={42}
          height={42}
          loading="lazy"
          className="dark:invert"
        />
      </div>

      <Drawer
        anchor="right"
        open={shouldShowDrawer}
        onClose={closeDrawer}
        onClick={closeDrawer}
      >
        <div className="flex flex-col pr-8 pl-4 pt-8 gap-4 dark:bg-slate-900 h-full min-w-[125px]">
          <BlogLinks />
        </div>
      </Drawer>
    </header>
  );
};
