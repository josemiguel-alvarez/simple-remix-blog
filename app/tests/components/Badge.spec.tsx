import { Badge } from "~/components/Badge";
import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

describe("Badge", () => {
  test("renders a link with a label", () => {
    render(<Badge label="Test Badge" linkTo="http://example.com" />, {
      wrapper: BrowserRouter,
    });

    expect(screen.getByText("Test Badge")).toBeDefined();
    expect(screen.getByRole("link")).toBeDefined();
  });
});
