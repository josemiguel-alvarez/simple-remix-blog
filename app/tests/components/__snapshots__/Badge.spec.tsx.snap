// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge renders correctly 1`] = `
<div>
  <div
    class="not-prose"
  >
    <a
      class="no-underline relative inline-block text-sm font-medium text-blue-700 group"
      href="/http://example.com"
    >
      <span
        class="absolute inset-0 transition-transform translate-x-0.5 translate-y-0.5 bg-blue-700 group-hover:translate-y-0 group-hover:translate-x-0"
      />
      <span
        class="relative block px-4 py-2 bg-white border border-current"
      >
        Test Badge
      </span>
    </a>
  </div>
</div>
`;
